### PR TITLE
Unify checks

### LINKED ISSUE #10821: Generalize internal checks for precision plugin type, training type, accelerator type
## Proposed refactor

Internally, our checks against the type of Accelerator, Precision type, strategy is not robust towards custom instances passed in by the user. 

### Motivation

Internally, some operations in the optimization, logging, etc. need a different code path depending on 1) Accelerator type (cpu, gpu) or 2) Precision type (apex, native) or 3) strategy type (ddp, ddp-spawn, ...). Currently we have this pattern: 

```py
if trainer._device_type == DeviceType.CPU:
    # do something only for cpu


if trainer._amp_backend == AMPType.Apex:
    # do something differently for apex
```

### Pitch

Change these to

```py
if isinstance(trainer.accelerator, CPUAccelerator):
    # do something only for cpu


if isinstance(trainer.precision_plugin, ApexPrecisionPlugin):
    # do something differently for apex
```

This has the benefits:
- User passes in custom plugins (subclasses of our plugins)
- Encapsulation: Protected members _device_type, _strategy_type, won't be abusively accessed publicly anymore. They remain an implementation detail of AcceleratorConnector
- Minimally simplifies AcceleratorConnector logic

### Additional context

Discusson started in https://github.com/PyTorchLightning/pytorch-lightning/pull/10596

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @justusschock @awaelchli @rohitgr7 @kaushikb11 @akihironitta @ananthsub

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?
fixes  #10821 

Generalizes Type-Checks

### Does your PR introduce any breaking changes? If yes, please list them.
Yes, deprecates `trainer.amp_backend` and `MixedPrecisionPlugin.backend`

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: pytorch_lightning/lite/lite.py
Comment (justusschock) [inline pytorch_lightning/lite/lite.py:91]: To Reviewers: I couldn't find any other way to validate this, but I think here it is fine to use the types, as this is for input validation and not for checking which mode we are in.

File: pytorch_lightning/lite/lite.py
Comment (justusschock) [reply pytorch_lightning/lite/lite.py:510]: let's give this a try. Haven't seen this syntax in that usecase 🤯

File: pytorch_lightning/lite/lite.py
Comment (justusschock) [inline pytorch_lightning/lite/lite.py:513]:
```suggestion
```

File: pytorch_lightning/lite/lite.py
Comment (carmocca) [inline pytorch_lightning/lite/lite.py:94]:
```suggestion
```

File: pytorch_lightning/lite/lite.py
Comment (carmocca) [inline pytorch_lightning/lite/lite.py:488]:
```suggestion
```

File: pytorch_lightning/lite/lite.py
Comment (carmocca) [inline pytorch_lightning/lite/lite.py:510]:
```suggestion
```

File: pytorch_lightning/lite/lite.py
Comment (carmocca) [inline pytorch_lightning/lite/lite.py:501]: Can't we remove this and `_supported_strategy_types` now?

File: pytorch_lightning/plugins/precision/mixed.py
Comment (carmocca) [inline pytorch_lightning/plugins/precision/mixed.py:25]: Why abstractmethod?

```suggestion
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/lite/lite.py
Comment (rohitgr7) [inline pytorch_lightning/lite/lite.py:483]:
```suggestion
                f"`accelerator={accelerator!r}` is not a valid choice for `LightningLite`."
```

File: pytorch_lightning/lite/lite.py
Comment (rohitgr7) [inline pytorch_lightning/lite/lite.py:492]:
```suggestion
                f"`strategy={strategy!r}` is not a valid choice for `LightningLite`."
```

File: pytorch_lightning/lite/lite.py
Comment (rohitgr7) [inline pytorch_lightning/lite/lite.py:512]: Bagua is not supported?

File: pytorch_lightning/plugins/precision/apex_amp.py
Comment (rohitgr7) [inline pytorch_lightning/plugins/precision/apex_amp.py:117]:
```suggestion
            "The backend property has been deprecated in v1.6 and will be removed in v1.7."
```

File: pytorch_lightning/plugins/precision/apex_amp.py
Comment (rohitgr7) [inline pytorch_lightning/plugins/precision/apex_amp.py:118]:
```suggestion
            " Please switch to `isinstance(X, ApexMixedPrecisionPlugin)` check instead."
```

File: pytorch_lightning/plugins/precision/mixed.py
Comment (rohitgr7) [inline pytorch_lightning/plugins/precision/mixed.py:29]:
```suggestion
        Switch to instance check instead
```

File: pytorch_lightning/plugins/precision/mixed.py
Comment (rohitgr7) [inline pytorch_lightning/plugins/precision/mixed.py:27]:
```suggestion
        """Deprecated in v1.6.
```
I think for sphinx, there is a format to represent the deprecated method.

File: pytorch_lightning/plugins/precision/native_amp.py
Comment (rohitgr7) [inline pytorch_lightning/plugins/precision/native_amp.py:136]:
```suggestion
            "The backend property has been deprecated in v1.6 and will be removed in v1.7."
            " Please switch to `isinstance(X, NativeMixedPrecisionPlugin)` check instead."
```

File: pytorch_lightning/trainer/trainer.py
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:2100]:
```suggestion
            "amp_backend is deprecated in v1.6 and will be removed in v1.7. "
            "Use `isinstance` check against the `PrecisionPlugins` directly."
```
are these following one deprecation cycle?

File: tests/deprecated_api/test_remove_1-7.py
Comment (rohitgr7) [inline tests/deprecated_api/test_remove_1-7.py:539]:
```suggestion
    with pytest.deprecated_call(match="amp_backend is deprecated in v1.6 and will be removed in v1.7."):
        trainer.amp_backend


def test_v1_7_0_mixed_precision_plugin_backend_native():
    plugin = NativeMixedPrecisionPlugin(16, "cpu")

    with pytest.deprecated_call(match="The backend property has been deprecated in v1.6 and will be removed in v1.7."):
        plugin.backend


@RunIf(amp_apex=True)
def test_v1_7_0_mixed_precision_plugin_backend_apex():
    plugin = ApexMixedPrecisionPlugin()
    with pytest.deprecated_call(match="The backend property has been deprecated in v1.6 and will be removed in v1.7."):
```

**Review by rohitgr7 — APPROVED** [review]
small comments.

File: pytorch_lightning/lite/lite.py
Comment (justusschock) [reply pytorch_lightning/lite/lite.py:94]: whoops

File: pytorch_lightning/lite/lite.py
Comment (justusschock) [reply pytorch_lightning/lite/lite.py:501]: short answer: no, See https://github.com/PyTorchLightning/pytorch-lightning/pull/12069#discussion_r815325126

long answer:
no, they are still necessary for input validation unfortunately. If this is removed, the input validation can only be done after the accelerator and strategy init and in case of unsupported arguments, the accelerator connector may error out with rather hard to interpret key errors or fall back to cpu in which case the user won't get any notification on this.

File: pytorch_lightning/plugins/precision/mixed.py
Comment (justusschock) [reply pytorch_lightning/plugins/precision/mixed.py:25]: because this should be implemented by all the MixedPrecisionPlugins. But you're right. Without ABC as base, this doesn't work :)

File: pytorch_lightning/lite/lite.py
Comment (justusschock) [reply pytorch_lightning/lite/lite.py:512]: I am not sure. It probably hasn't been tested yet. I kept these as matches from the _StrategyTypes @awaelchli put there initially.

**Comment by rohitgr7** [comment]
are we following one release cycle for deprecation here?

**Comment by justusschock** [comment]
I would say so. Mainly because they were part of the accelerator-strategy API that has not yet been stable.

File: pytorch_lightning/plugins/precision/mixed.py
Comment (Borda) [inline pytorch_lightning/plugins/precision/mixed.py:26]: lest use `.. deprecated::`

File: pytorch_lightning/trainer/connectors/data_connector.py
Comment (four4fish) [inline pytorch_lightning/trainer/connectors/data_connector.py:167]: Deepspeed could use amp, but it's not extended from MixedPrecisionPlugin
@daniellepintz's change should fix this https://github.com/PyTorchLightning/pytorch-lightning/pull/12323

File: pytorch_lightning/trainer/trainer.py
Comment (daniellepintz) [inline pytorch_lightning/trainer/trainer.py:2088]: Don't forget to add this to the changelog :)

**Review by daniellepintz — APPROVED** [review]
LGTM once comments are addressed

File: pytorch_lightning/plugins/precision/mixed.py
Comment (rohitgr7) [inline pytorch_lightning/plugins/precision/mixed.py:25]: some issue with docstrings here :)

Comment (justusschock) [reply pytorch_lightning/plugins/precision/mixed.py]: yes, just fixed that. pre-commit kept reformatting these to the format sphinx could not read -.-

**Review by awaelchli — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/strategies/deepspeed.py
Comment (awaelchli) [inline pytorch_lightning/strategies/deepspeed.py:656]: Note, the test (test_trainer_model_hook_system_fit) failed because the type was actually `DeepSpeedPrecisionPlugin` which also has a field amp_type (set to NATIVE).
The same would be true for the apex check in the "elif" block below. 
Both if statements will be skipped with the changes in this PR. 

Since the DeepSpeedPrecisionPlugin is a weird basically bundling both native and apex together, we probably have to revert the changes here or have a more radical change.

File: pytorch_lightning/lite/lite.py
Comment (kaushikb11) [reply pytorch_lightning/lite/lite.py:509]: cc: @awaelchli

Comment (awaelchli) [reply pytorch_lightning/lite/lite.py:509]: Yes, in theory Lite would support it since it is a modification of the DDPStrategy. However, we should sanity check this first before we add it to the types here. I created an issue for this: #12500

File: pytorch_lightning/strategies/deepspeed.py
Comment (justusschock) [reply pytorch_lightning/strategies/deepspeed.py:656]: Deepspeed only supports native AMP, right?

Comment (awaelchli) [reply pytorch_lightning/strategies/deepspeed.py:656]: It appears so yes. I too think it would be worth investigating  #12323

**Review by Borda — COMMENTED** [review]
What is the status of this PR, @justusschock?

**Comment by justusschock** [comment]
I'll continue this once I'm back from vacation

**Comment by carmocca** [comment]
Good. This is now blocked by the Lite changes anyways

**Comment by Borda** [comment]
@justusschock @awaelchli, how it is going here? :otter:

**Comment by justusschock** [comment]
This will be on-hold until the standalone lite has been completed. Then we can further pursue this

**Comment by Borda** [comment]
> This will be on-hold until the standalone lite has been completed. Then we can further pursue this

I think it is done there :)

**Comment by carmocca** [comment]
The Lite code has been copied over but PL's abstractions haven't fully integrated into them yet. So we should still wait

**Pull request closed without merging**
